packages/contracts/src/2.0.0/forwarder/MixinForwarderCore.sol

```diff
@@ -53,377 +48,257 @@ contract MixinForwarderCore is
+    /// @param orders Array of order specifications containing ZRX as makerAsset and WETH as takerAsset.
+    /// @param zrxBuyAmount Desired amount of ZRX to buy.
+    /// @param signatures Proofs that orders have been created by makers.
+    /// @return totalFillResults Amounts filled and fees paid by maker and taker.
+    function marketBuyZrx(
         LibOrder.Order[] memory orders,
-        bytes[] memory signatures,
-        LibOrder.Order[] memory feeOrders,
-        bytes[] memory feeSignatures,
-        uint256 makerTokenFillAmount
+        uint256 zrxBuyAmount,
+        bytes[] memory signatures
     )
         internal
-        returns (LibFillResults.FillResults memory totalFillResults)
+        returns (FillResults memory totalFillResults)
     {
-        // We read the maker token address to check if it is ZRX and later use it for transfer
-        address makerTokenAddress = LibBytes.readAddress(orders[0].makerAssetData, 16);
-        // We assume that asset being bought by taker is the same for each order.
-        // Rather than passing this in as calldata, we copy the makerAssetData from the first order onto all later orders.
-        orders[0].takerAssetData = WETH_ASSET_DATA;
-        // We can short cut here for effeciency and use buyFeeTokensInternal if maker asset token is ZRX
-        // this buys us exactly that amount taking into account the fees. This saves gas and calculates the rate correctly
-        FillResults memory marketBuyResults;
-        if (makerTokenAddress == address(ZRX_TOKEN)) {
-            marketBuyResults = marketBuyZrxInternal(
-                orders,
-                signatures,
-                makerTokenFillAmount
-            );
-            // When buying ZRX we round up which can result in a small margin excess
-            require(
-                marketBuyResults.makerAssetFilledAmount >= makerTokenFillAmount,
-                "UNACCEPTABLE_THRESHOLD"
-            );
-            addFillResults(totalFillResults, marketBuyResults);
-            require(
-                isAcceptableThreshold(
-                    safeAdd(totalFillResults.makerAssetFilledAmount, totalFillResults.takerFeePaid), // Total ZRX
-                    totalFillResults.makerAssetFilledAmount // amount going to msg.sender
-                ),
-                "UNACCEPTABLE_THRESHOLD"
-            );
-        } else {
-            FillResults memory calculatedMarketBuyResults = calculateMarketBuyResults(orders, makerTokenFillAmount);
-            if (calculatedMarketBuyResults.takerFeePaid > 0) {
-                // Fees are required for these orders. Buy enough ZRX to cover the future market buy
-                FillResults memory zrxMarketBuyResults = marketBuyZrxInternal(
-                    feeOrders,
-                    feeSignatures,
-                    calculatedMarketBuyResults.takerFeePaid
+        // Do nothing if zrxBuyAmount == 0
+        if (zrxBuyAmount > 0) {
```

Comment:
Nit - this block feels kind of long to indent.
Could instead do `if (zrxBuyAmount == 0) return totalFillResults;`